File: ECE39500/exams/midterm-2/midterm-2/PartA/Compiler.py
from CompilerState import CompilerState
from StateFactory import StateFactory
from AddState import AddState
from CVCVState import CVCVState
from InitState import InitState
from EofState import EofState
from ErrorState import ErrorState
from VCVState import VCVState
import sys
import re

class Compiler():
    INT_MIN = -sys.maxint - 1

    def __init__(self):
        self.DEBUG = False
        self.valueCount = -1
        self.values = (INT_MIN, INT_MIN, INT_MIN)
        self.targetState = None
        self.state = None
        return

    def compile(self, fileName):
        self.state = StateFactory.getState("InitState")
        with open(filename, "r") as f:
            lines = f.readlines()
        for line in lines:
            self.parseline(line)


    def parseline(self, line):
        if self.DEBUG:
            print(f"\nParsing input line: {line}")
        line = re.sub(r",", " , ", line)
        line = re.sub(r"\s+", " ", line)
        tokens = line.split(" ")
        for token in tokens:
            self.takeAction(token)
        self.state.eol()
        return
    
    def takeAction(token):
        commands = ("addi", "ld0", "add", "print", "exit")
        if token.isnumeric():
            self.state.value(token)
        elif token in commands:
            self.values[0] = INT_MIN
            self.values[1] = INT_MIN
            self.values[2] = INT_MIN
            self.valueCount = -1
            self.state.value(token)
        elif token == ",":
            self.state.comma()
        else:
            print(f"Unknown token :: {token} ::")
            self.state = StateFactory.getState("ErrorState")

File: ECE39500/exams/midterm-2/PartB/src/Compiler.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

public class Compiler {

   // ld0 regi
   // addi regi, int
   // add regi, regj, regk // regi = regj + regk
   // print regi
   // halt

   public static final boolean DEBUG = false;

   public static int valueCount = -1;
   public static int[ ] values = {Integer.MIN_VALUE, Integer.MIN_VALUE, 
                                  Integer.MIN_VALUE};

   public static CompilerState targetState = null;
   public static CompilerState state = null;

   public Compiler( ) { }

   public void compile(String fileName) {
      state = StateFactory.getState("InitState");
      BufferedReader reader = null;
      try {
         reader = new BufferedReader(new FileReader(fileName));
         String line = reader.readLine();
         while (line != null) {
            parseLine(line);
            line = reader.readLine();
         }
         state.eof( ); // CHANGED
      } catch (IOException e) {
         e.printStackTrace();
      } finally {
         try {
            if (reader == null) {
               System.out.println("file "+fileName+" not opened");
            } else {
               reader.close();
            }
         } catch (IOException e) {e.printStackTrace( );}
      }
   }

   private void parseLine(String line) {
      if (DEBUG) {
         System.out.println("\nParsing input line: "+line);
      }

      line = line.trim( );
      line = line.replaceAll(",", " , ");
      line = line.replaceAll("\\s+", " ");
      String[ ] tokens = line.split("\\s");
      
      for (int i = 0; i < tokens.length; i++) {
         takeAction(tokens[i]); 
      }
      state.eol( );
   }

   private void takeAction(String token) {
      if (DEBUG) {
         System.out.println("Taking action with state "+state);
      }
      if (token.matches("[+-]?[0-9][0-9]*")) { 
        state.value(token);
      } else if (token.matches("addi|ld0|add|print|exit|store")) {
         for (int i = 0; i < 3; i++) {
            values[i] = Integer.MIN_VALUE;
         }
         valueCount = -1;
         state.value(token);
      } else if (token.equals(",")) {
         state.comma( );
      } else {
         System.out.println("Unknown token ::" + token+"::"); 
         state = StateFactory.getState("ErrorState");
      }
   }
}
